currencies/rate_converter.go

```diff
@@ -0,0 +1,165 @@
+	rates                 *Rates
+	lastUpdated           *time.Time
+}
+
+// NewRateConverter returns a new RateConverter
+func NewRateConverter(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+) *RateConverter {
+	return &RateConverter{
+		httpClient:            httpClient,
+		updates:               make(chan *Rates),
+		done:                  make(chan bool),
+		fetchingInterval:      fetchingInterval,
+		retryFetchingInterval: retryFetchingInterval,
+		ticker:                nil,
+		syncSourceURL:         syncSourceURL,
+		rates:                 nil,
+	}
+}
+
+// fetch allows to retrieve the currencies rates from the syncSourceURL provided
+func (rc *RateConverter) fetch() (*Rates, error) {
+	response, err := rc.httpClient.Get(rc.syncSourceURL)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	updatedRates := &Rates{}
+	err = json.NewDecoder(response.Body).Decode(&updatedRates)
+	if err != nil {
+		return nil, err
+	}
+	updatedRates.ComputeDataAsOf()
+
+	return updatedRates, err
+}
+
+// updateRates allows to update the internal currencies rates map with the one provided
+func (rc *RateConverter) updateRates(updatedRates *Rates) {
+	rc.Lock()
+	defer rc.Unlock()
+	rc.rates = updatedRates
+	now := time.Now()
+	rc.lastUpdated = &now
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() (*Rates, error) {
+	rates, err := rc.fetch()
+	if err == nil {
+		rc.Lock()
+		defer rc.Unlock()
+		rc.rates = rates
+		now := time.Now()
+		rc.lastUpdated = &now
+	}
+
+	return rates, err
+}
+
+// StartPeriodicFetching starts the periodic fetching at the given interval
+// triggers a first fetch when called before the first tick happen in order to initialize currencies rates map
+// returns a chan in which the number of data updates everytime a new update was done
+func (rc *RateConverter) StartPeriodicFetching() (chan int, error) {
+	rc.ticker = time.NewTicker(rc.fetchingInterval)
+	updatesTicksCount := 0
+	updatesNotifier := make(chan int)
+
+	go func() {
+		for {
+			select {
+			case <-rc.ticker.C:
+				go func() {
+					rates, err := rc.fetch()
+					if err == nil {
+						rc.updates <- rates
+					}
+					// TODO: Handle retry strategy
+				}()
+			case rates := <-rc.updates:
+				go func() {
+					rc.updateRates(rates)
+					updatesTicksCount++
+					updatesNotifier <- updatesTicksCount
+				}()
+			case <-rc.done:
+				if rc.ticker != nil {
+					rc.ticker.Stop()
+					rc.ticker = nil
+				}
+				return
+			default:
+			}
+		}
+	}()
+
+	// Trigger a first tick at the very begining
+	// the ticker will tick later
+	rates, err := rc.fetch()
+	if err != nil {
+		return nil, err
+	}
+	rc.updateRates(rates)
+
+	return updatesNotifier, nil
+}
+
+// StopPeriodicFetching stops the periodic fetching while keeping the latest currencies rates map
+func (rc *RateConverter) StopPeriodicFetching() {
+	rc.done <- true
```

Comment:
What's the purpose for this?

Seems like we'd start fetching rates when the app starts, and continue until the server shuts down.

Do you plan to use this to for graceful shutdown? If so, you'd need another channel which blocks until the goroutine from `StartPeriodicFetching` totally returns.

Reply:
We might not need it indeed. From your previous comment, If we don't need the start, we might not need the Stop.
But I still think that for convenience, we should still return a chan in which we send ticks. So if the fetching starts directly on object construction, then, we can return this chan in `rateConverter, tickerChan, err := NewRateConverter(...)` what do you think ?

Also, if we want to align on the `time.ticker` implementation, the later does have the `Stop()` function, so maybe good to keep.

Reply:
I don't have a strong opinion... but I'd lean towards leaving it in and fixing it. Graceful shutdowns are nice to have. For this use-case, though, I'd call them very low priority.

If you keep it, my main concern is that it returns "too early" to be used as a graceful shutdown function. For example: imagine it gets called while the `go func()` inside `StartPeriodicFetching` is in the middle of making an HTTP call/rate update.

This function would send a signal on `rc.done`... but it will return before the `go func()` ever finishes its call and processes the "done" signal. So there's no real way for the caller to know when the shutdown has truly finished.

> So if the fetching starts directly on object construction, then, we can return this chan in rateConverter, tickerChan, err := NewRateConverter(...) what do you think ?

That's an option... my main concern is that channels have a limited buffer space. If callers don't read the values from that `tickerChan`, it will eventually fill up and block future updates from happening.

Is there a use-case for "listening" on updates? One option is to make callers pass in the channel:

```
func NewRateConverter(func NewRateConverter(
  httpClient httpClient,
  syncSourceURL string,
  updater <-chan {any},
  ...
) *RateConverter
```

Callers could just send in a `NewTicker().C` for `updater`, and it would replace your current `fetchingInterval`. If the caller also wants to be alerted of updates, they could [fan out](https://blog.golang.org/pipelines) the events from a ticker, pass one into here, and listen to the other themselves.

Reply:
Done, I allowed clients to pass a chan in which they will be listening notifications.
There are now two functions `NewRateConverter` and `NewRateConverterWithNotifier` to cronstruct new `RateConverter`. By default, `NewRateConverter` doesn't track updates whereas `NewRateConverterWithNotifier` will, so clients can pick the one they want and will also avoid having a forgotten chan provided by the client.

Reply:
Done, I allowed clients to pass a chan in which they will be listening notifications.
There are now two functions NewRateConverter and NewRateConverterWithNotifier to cronstruct new RateConverter. By default, NewRateConverter doesn't track updates whereas NewRateConverterWithNotifier will, so clients can pick the one they want and will also avoid having a forgotten chan provided by the client.

Reply:
If `StopPeriodicFetching` is kept this should be changed to do to a close to ensure all goroutines are broadcasted to.
```go
close(rc.done)
```
Even if there is currently only one goroutine it will help protect future work if more are added or if `startPeriodicFetching` is called more than once.

Reply:
Good point :) 
Done